wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj2/command/Resource.java

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package edu.wpi.first.wpilibj2.command;
+
+import java.util.Set;
+import java.util.function.Supplier;
+
+/**
+ * `Resource` is the <a href="https://en.wikipedia.org/wiki/Lock_(computer_science)">locking</a> 
+ * implementation for the Command-based concurrency model, encapsulating program state that should 
+ * only be controlled by one {@link Command} at a time.  If multiple commands requiring the same
+ * resource are scheduled at the same time, the {@link CommandScheduler} will resolve the resource
+ * contention by interrupting (or refusing to schedule) all but one of the commands.
+ * 
+ * <p>Typically, protected state represents hardware resources (like motors) for which control
+ * disputes would present an operational hazard.  But, it may also be appropriate to
+ * protect software resources, like vision pipelines, whose outputs might become invalid if
+ * their configurations are changed inappropriately during a task.
+ * 
+ * <p>Classes containing protected state should implement this interface.  Commands that require
+ * said state should be written using the Command factories (e.g. {@link Resource#run(Runnable)}
+ * defined on this interface, which automatically declare the resource requirement.
+ */
+public interface Resource {
+  /**
+   * This method is called periodically by the {@link CommandScheduler}. Useful for updating
+   * resource-specific state that you don't want to offload to a {@link Command}. Teams should try
+   * to be consistent within their own codebases about which responsibilities will be handled by
+   * Commands, and which will be handled here.
+   */
+  default void periodic() {}
+
+  /**
+   * This method is called periodically by the {@link CommandScheduler}. Useful for updating
+   * resource-specific state that needs to be maintained for simulations, such as for updating
+   * {@link edu.wpi.first.wpilibj.simulation} classes and setting simulated sensor readings.
+   */
+  default void simulationPeriodic() {}
```

Comment:
These seem more appropriate for subsystems than for a more generic "resource" type. In my mind, resources are really just for the ownership model for commands, not periodic things in and of themselves. Periodic code with a resource can be done just by calling the `periodic()` function in `robotPeriodic()`, and the same for the simulation periodic function. Though I _have_ had students tell me that "it's just easier" to use a subsystem to get access to a periodic method that runs automatically, instead of them writing the one line of code themselves :shrug: